ambry-store/src/main/java/com.github.ambry.store/StoreMessageReadSet.java

```diff
@@ -13,96 +13,127 @@
  */
 package com.github.ambry.store;
 
+import com.github.ambry.utils.Pair;
+import java.io.Closeable;
 import java.io.DataInputStream;
-import java.nio.ByteBuffer;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
 import java.io.File;
 import java.io.IOException;
+import java.nio.ByteBuffer;
 import java.nio.channels.FileChannel;
 import java.nio.channels.WritableByteChannel;
 import java.util.Collections;
 import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 
 /**
  * A read option class that maintains the offset and size
  */
-class BlobReadOptions implements Comparable<BlobReadOptions> {
-  private final Long offset;
+class BlobReadOptions implements Comparable<BlobReadOptions>, Closeable {
+  private final LogSegment segment;
+  private final Pair<File, FileChannel> segmentView;
+  private final Offset offset;
   private final Long size;
   private final Long ttl;
   private final StoreKey storeKey;
-  private Logger logger = LoggerFactory.getLogger(getClass());
+  private final AtomicBoolean open = new AtomicBoolean(true);
+  private final Logger logger = LoggerFactory.getLogger(getClass());
 
   private static final short version = 0;
   private static final short Version_Length = 2;
   private static final short Offset_Length = 8;
   private static final short Size_Length = 8;
   private static final short TTL_Length = 8;
 
-  BlobReadOptions(long offset, long size, long ttl, StoreKey storeKey) {
+  BlobReadOptions(Log log, Offset offset, long size, long ttl, StoreKey storeKey) {
+    segment = log.getSegment(offset.getName());
```

Comment:
That's fine - that would be a programming error i.e the index had an entry that referred to no log segment. In that case, `NullPointerException` is as good as any. 

Did you have something else in mind?